Mathlib/Topology/Algebra/Module/Action.lean

```diff
@@ -0,0 +1,238 @@
+that the action topology also has these properties. -/
+abbrev actionTopology : TopologicalSpace A :=
+  sInf {t | @ContinuousSMul R A _ _ t ∧ @ContinuousAdd A t _}
+
+/-- A class asserting that the topology on a module over a topological ring `R` is
+the action topology. See `actionTopology` for more discussion of the action topology. -/
+class IsActionTopology [τA : TopologicalSpace A] : Prop where
+  eq_actionTopology' : τA = actionTopology R A
+
+theorem eq_ActionTopology [τA : TopologicalSpace A] [IsActionTopology R A] :
+    τA = actionTopology R A :=
+  IsActionTopology.eq_actionTopology' (R := R) (A := A)
+
+/-- Scalar multiplication `• : R × A → A` is continuous if `R` is a topological
+ring, and `A` is an `R` module with the action topology. -/
+theorem ActionTopology.continuousSMul : @ContinuousSMul R A _ _ (actionTopology R A) :=
+  /- Proof: We need to prove that the product topology is finer than the pullback
+     of the action topology. But the action topology is an Inf and thus a limit,
+     and pullback is a right adjoint, so it preserves limits.
+     We must thus show that the product topology is finer than an Inf, so it suffices
+     to show it's a lower bound, which is not hard. All this is wrapped into
+     `continuousSMul_sInf`.
+  -/
+  continuousSMul_sInf <| fun _ _ ↦ by simp_all only [Set.mem_setOf_eq]
+
+/-- Addition `+ : A × A → A` is continuous if `R` is a topological
+ring, and `A` is an `R` module with the action topology. -/
+theorem ActionTopology.continuousAdd : @ContinuousAdd A (actionTopology R A) _ :=
+  continuousAdd_sInf <| fun _ _ ↦ by simp_all only [Set.mem_setOf_eq]
+
+instance IsActionTopology.toContinuousSMul [TopologicalSpace A] [IsActionTopology R A] :
+    ContinuousSMul R A := eq_ActionTopology R A ▸ ActionTopology.continuousSMul R A
+
+-- this can't be an instance because typclass inference can't be expected to find `R`.
+theorem IsActionTopology.toContinuousAdd [TopologicalSpace A] [IsActionTopology R A] :
+    ContinuousAdd A := eq_ActionTopology R A ▸ ActionTopology.continuousAdd R A
+
+/-- The action topology is `≤` any topology making `A` into a topological module. -/
+theorem actionTopology_le [τA : TopologicalSpace A] [ContinuousSMul R A] [ContinuousAdd A] :
+    actionTopology R A ≤ τA := sInf_le ⟨inferInstance, inferInstance⟩
+
+end basics
+
+namespace ActionTopology
+
+section zero
+
+instance instSubsingleton (R : Type*) [TopologicalSpace R] (A : Type*) [Add A] [SMul R A]
+    [Subsingleton A] [TopologicalSpace A] : IsActionTopology R A where
+  eq_actionTopology' := by
+    ext U
+    simp only [isOpen_discrete]
+
+end zero
+
+section one
+
+/-
+
+We now fix once and for all a topological semiring `R`.
+
+We first prove that the action topology on `R` considered as a module over itself,
+is `R`'s topology.
+
+-/
+instance instSelf (R : Type*) [Semiring R] [τR : TopologicalSpace R] [TopologicalSemiring R] :
```

Comment:
```suggestion
instance _root_.TopologicalSemiring.toIsActionTopology (R : Type*) [Semiring R] [τR : TopologicalSpace R] [TopologicalSemiring R] :
```
can you add the `MulOpposite` action too?

Reply:
Do you mean "prove `IsActionTopology R Rᵐᵒᵖ`? Or more generally "prove IsActionTopology R M -> IsActionTopology R Mᵐᵒᵖ`? The simplest way to do this would be to deduce it from the fact that `M` and  `Mᵐᵒᵖ` are ContinuousLinearEquiv, and then it follows immediately from the results in this PR. However we don't seem to have this result; we have that they're isomorphic as R-modules and that they're homeomorphic, but we don't have the `ContinuousLinearEquiv`. Do you want me to make a new file adding only this result? In this PR? Or did I misunderstand?

Reply:
I think you are misunderstanding. Eric is not talking about opposing the `M` but opposing the `R` instead

Reply:
... and this really cannot be deduced from the `R R` version because the action is different (and incompatible with the identification of `R` with `R^mop`)

In general, if you can talk about `a * b`, then you can try talking about `a • b` or about `op b • a`. Currently, you gave done the first but not the second.

Reply:
If this is what Eric means then my comments above still apply, you just replace R with Rᵐᵒᵖ. So if my current understanding of what I'm being asked to do is correct, one solution would be to make a new file containing the one result

```lean
import Mathlib.Algebra.Module.Opposites
import Mathlib.Topology.Algebra.Module.Basic

def MulOpposite.opContinuousLinearEquiv : R ≃L[Rᵐᵒᵖ] Rᵐᵒᵖ where
  __ := (MulOpposite.opLinearEquiv Rᵐᵒᵖ)
```

(because that surely doesn't go in this file), and then in this file we add

```
instance mulOpposite (R : Type*) [Semiring R] [TopologicalSpace R] [TopologicalSemiring R] :
    IsActionTopology Rᵐᵒᵖ R := .iso (MulOpposite.opContinuousLinearEquiv R).symm
```

An alternative approach is to put `MulOpposite.opContinuousLinearEquiv` into `Mathlib.Topology.Algebra.Module.Basic` and add `import Mathlib.Algebra.Module.Opposites` to that file.

Thoughts?

Reply:
But surely `R` isn't `R`-linearly isomorphic to `Rᵐᵒᵖ`?

Reply:
Do we secretly have a diamond where `op a • op b` is both `op (b * a)` and `op (a * b)` depending on the instance you're inferring?

Reply:
I slightly suspect [`MulOpposite.instMulAction`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Action/Opposite.html#MulOpposite.instMulAction) defines `a • op b := op (a • b)` is the issue

Reply:
I'm not really following this discussion, but I pushed a commit with the sorry filled.

Reply:
> But surely `R` isn't `R`-linearly isomorphic to `Rᵐᵒᵖ`?

https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Opposites.html#MulOpposite.opLinearEquiv

Can you clarify what you're worried about here?

Reply:
I guess so yeah. I was working from the assumption that `R` would act on `Rᵐᵒᵖ` on the right, but that seems to be incorrect. All I'm saying is that `R` and `Rᵐᵒᵖ` are different as `Rᵐᵒᵖ`-modules (and should also be different as `R `-modules, but seem to be the same on current mathlib)

Reply:
Just to be clear, in the context of "self-actions", the only two we care about are `Rᵐᵒᵖ` acting on `R`, and `R` acting on `R`. This was the only case I was talking about originally, and these should usually be on equal footing.

Reply:
> All I'm saying is that R and Rᵐᵒᵖ are different as Rᵐᵒᵖ-modules 

This is false? Send `x` to `op x` and show `Rᵐᵒᵖ`-linearity by considering actions by `op r`: `op (op r • x) = op (x * r) = op r * op x = op r • op x`

Reply:
> An alternative approach is to put `MulOpposite.opContinuousLinearEquiv` into `Mathlib.Topology.Algebra.Module.Basic` and add `import Mathlib.Algebra.Module.Opposites` to that file (of length >= 2400 lines).

This sounds good to me.

Reply:
Here is what we currently have:
```lean
import Mathlib

variable (R : Type*) [Semiring R]

open MulOpposite

example (a b : R) : a • b = a * b := rfl
example (a b : R) : (op a) • b = b * a := rfl
example (a b : R) : (op a) • (op b) = op (b * a) := rfl
example (a b : R) : a • (op b) = op (a * b) := rfl
```

I think I've got to the bottom of the confusion (Yael might have also done this). Yael is observing that in a parallel universe mathlib could have `a • (op b) = op (b * a)` because if someone asked for a natural left action of `R` on `Rᵐᵒᵖ` then one very reasonable response would be "well there's a natural right action of `R` on `R` so let's use that". This would break the symmetries in the above examples though. Indeed I would have *guessed* that `Rᵐᵒᵖ` and `R` would be different as `R`-modules (they're certainly not in general isomorphic as rings). However mathlib has made other decisions here.

Reply:
> > An alternative approach is to put `MulOpposite.opContinuousLinearEquiv` into `Mathlib.Topology.Algebra.Module.Basic` and add `import Mathlib.Algebra.Module.Opposites` to that file (of length >= 2400 lines).
> 
> This sounds good to me.

OK done. I've also (in another commit) changed the name to "module topology" following Yury's suggestion.

Reply:
> > All I'm saying is that R and Rᵐᵒᵖ are different as Rᵐᵒᵖ-modules 
> 
> This is false? Send `x` to `op x` and show `Rᵐᵒᵖ`-linearity by considering actions by `op r`: `op (op r • x) = op (x * r) = op r * op x = op r • op x`
> 

I am now confused because I was dead certain one could not deduce the `R Rᵐᵒᵖ` result from the `R R` one

Reply:
> > All I'm saying is that R and Rᵐᵒᵖ are different as Rᵐᵒᵖ-modules 
> 
> This is false? Send `x` to `op x` and show `Rᵐᵒᵖ`-linearity by considering actions by `op r`: `op (op r • x) = op (x * r) = op r * op x = op r • op x`
> 

I was confused because I was dead certain one could not deduce the `R Rᵐᵒᵖ` result from the `R R` one. I guess one actually can because the multiplication on `M` doesn't matter.

Reply:
Yes that's the point: if R acts on M then the R-action on M^mop in mathlib is defined to be the R-action on M because in this generality there's no other choice. But when M=R there is a second choice because R acts naturally both on the left and right of M in this case.

Reply:
Yes absolutely. I am happy to see it resolved.